app/controllers/application_controller.rb

```diff
@@ -5,6 +5,121 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "secret"
   end
 
-end
\ No newline at end of file
+  get '/' do
+    erb :'homepage'
+  end
+
+  get '/signup'  do
+    if !current_user
+      erb :'signup'
+    else
+      redirect to '/tweets'
+    end
+  end
+
+  post '/signup' do
+    if params[:username].empty? || params[:email].empty? || params[:password].empty?
+      redirect to '/signup'
+    else
+      @user = User.create(params)
+      session[:id] = @user.id
+      redirect to '/tweets'
+    end
+  end
+
+  get '/tweets' do
+    if current_user
+      erb :'tweets'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/login' do
+    if !current_user
+      erb :'login'
+    else
+      redirect to "/tweets"
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+
+      session[:id] = user.id
+      redirect to '/tweets'
+    else
+      redirect to "/tweets"
+    end
+  end
+
+  get '/logout' do
+    session.clear if current_user
+    redirect to '/login'
+  end
+
+  get '/tweets/new' do
+    if current_user
+      erb :'/tweets/new'
+    else
+      redirect to '/login'
+    end
+  end
+
+  post '/tweets/new' do
+    if params[:content].empty?
+      redirect to "/tweets/new"
+    else
+      @tweet = Tweet.create(params)
+      if current_user
+        @tweet.user = current_user
+        @tweet.save
+        redirect to "/tweets/#{@tweet.id}"
+      else
+        redirect to '/login'
+      end
+    end
+  end
+
+  get '/tweets/:id' do
+    if current_user
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/show'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if current_user
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/edit'
+    else
+      redirect to '/login'
+    end
+  end
+
+  post '/tweets/:id/edit' do
+    @tweet = Tweet.find(params[:id])
+    @tweet.update(content: params[:content])
+    erb :'tweets/show'
+  end
+
+  delete '/tweets/:id/delete' do
+    @tweet = Tweet.find(params[:id])
+    if current_user == @tweet.user
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.